File: frontend/src/utils/CustomCipher.ts
export class CustomCipher {
    private static ANILLO_1 = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    private static ANILLO_2 = "abcdefghijklmnopqrstuvwxyz";
    private static ANILLO_3 = "0123456789.,!@#$%^&*()_+-=";
    private static TODOS = CustomCipher.ANILLO_1 + CustomCipher.ANILLO_2 + CustomCipher.ANILLO_3;
    
    private static SECRET_KEY = 5;

    public static cifrar(texto: string): string {
        if (!texto) return texto;
        
        const k = this.SECRET_KEY;
        let texto_cifrado = "";
        const indice_oculto = ((k % 8) + 8) % 8;

        for (let i = 0; i < texto.length; i++) {
            const charItem = texto[i];

            if (charItem === ' ') {
                texto_cifrado += ' ';
                continue;
            }

            let char_cifrado = '';

            if (this.ANILLO_1.indexOf(charItem) !== -1) {
                const p = this.ANILLO_1.indexOf(charItem);
                const c = ((p + k) % 26 + 26) % 26;
                char_cifrado = this.ANILLO_1[c];
            } else if (this.ANILLO_2.indexOf(charItem) !== -1) {
                const p = this.ANILLO_2.indexOf(charItem);
                const c = ((p - k) % 26 + 26) % 26;
                char_cifrado = this.ANILLO_2[c];
            } else if (this.ANILLO_3.indexOf(charItem) !== -1) {
                const p = this.ANILLO_3.indexOf(charItem);
                const c = ((p + k) % 26 + 26) % 26;
                char_cifrado = this.ANILLO_3[c];
            } else {
                texto_cifrado += charItem;
                continue;
            }

            let bloque = "";
            for (let j = 0; j < 8; j++) {
                if (j === indice_oculto) {
                    bloque += char_cifrado;
                } else {
                    const randomChar = this.TODOS[Math.floor(Math.random() * this.TODOS.length)];
                    bloque += randomChar;
                }
            }
            texto_cifrado += bloque;
        }

        return texto_cifrado;
    }

    public static descifrar(texto_cifrado: string): string {
        if (!texto_cifrado) return texto_cifrado;
        
        const k = this.SECRET_KEY;
        let texto_descifrado = "";
        const indice_oculto = ((k % 8) + 8) % 8;

        let i = 0;
        while (i < texto_cifrado.length) {
            const charActual = texto_cifrado[i];

            if (charActual === ' ') {
                texto_descifrado += ' ';
                i++;
                continue;
            }

            if (this.TODOS.indexOf(charActual) === -1) {
                texto_descifrado += charActual;
                i++;
                continue;
            }

            if (i + 8 > texto_cifrado.length) {
                break;
            }

            const bloque = texto_cifrado.substring(i, i + 8);
            const char_cifrado = bloque[indice_oculto];

            if (this.ANILLO_1.indexOf(char_cifrado) !== -1) {
                const c = this.ANILLO_1.indexOf(char_cifrado);
                const p = ((c - k) % 26 + 26) % 26;
                texto_descifrado += this.ANILLO_1[p];
            } else if (this.ANILLO_2.indexOf(char_cifrado) !== -1) {
                const c = this.ANILLO_2.indexOf(char_cifrado);
                const p = ((c + k) % 26 + 26) % 26;
                texto_descifrado += this.ANILLO_2[p];
            } else if (this.ANILLO_3.indexOf(char_cifrado) !== -1) {
                const c = this.ANILLO_3.indexOf(char_cifrado);
                const p = ((c - k) % 26 + 26) % 26;
                texto_descifrado += this.ANILLO_3[p];
            }

            i += 8;
        }

        return texto_descifrado;
    }
}

File: backend/src/main/java/kentakitos/backend/util/CustomCipher.java
package kentakitos.backend.util;

import java.security.SecureRandom;

public class CustomCipher {

    private static final String ANILLO_1 = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    private static final String ANILLO_2 = "abcdefghijklmnopqrstuvwxyz";
    private static final String ANILLO_3 = "0123456789.,!@#$%^&*()_+-=";
    private static final String TODOS_LOS_CARACTERES = ANILLO_1 + ANILLO_2 + ANILLO_3;
    private static final SecureRandom random = new SecureRandom();

    /**
     * Cifra un texto utilizando el algoritmo de 'engranajes' (expansión a bloques de 8 con ruido).
     */
    public static String cifrar(String texto, int k) {
        if (texto == null) return null;
        StringBuilder texto_cifrado = new StringBuilder();
        
        // En Java el operador % puede dar números negativos, por lo que aplicamos: ((k % 8) + 8) % 8
        // para asegurar el mismo comportamiento matemático que Python.
        int indice_oculto = ((k % 8) + 8) % 8;

        for (char charItem : texto.toCharArray()) {
            if (charItem == ' ') {
                texto_cifrado.append(' ');
                continue;
            }

            char char_cifrado = '\0';

            if (ANILLO_1.indexOf(charItem) != -1) {
                int p = ANILLO_1.indexOf(charItem);
                int c = ((p + k) % 26 + 26) % 26;
                char_cifrado = ANILLO_1.charAt(c);
            } else if (ANILLO_2.indexOf(charItem) != -1) {
                int p = ANILLO_2.indexOf(charItem);
                int c = ((p - k) % 26 + 26) % 26;
                char_cifrado = ANILLO_2.charAt(c);
            } else if (ANILLO_3.indexOf(charItem) != -1) {
                int p = ANILLO_3.indexOf(charItem);
                int c = ((p + k) % 26 + 26) % 26;
                char_cifrado = ANILLO_3.charAt(c);
            } else {
                texto_cifrado.append(charItem);
                continue;
            }

            StringBuilder bloque = new StringBuilder();
            for (int i = 0; i < 8; i++) {
                if (i == indice_oculto) {
                    bloque.append(char_cifrado);
                } else {
                    bloque.append(TODOS_LOS_CARACTERES.charAt(random.nextInt(TODOS_LOS_CARACTERES.length())));
                }
            }
            texto_cifrado.append(bloque.toString());
        }
        return texto_cifrado.toString();
    }

    /**
     * Descifra un texto que fue encriptado con el algoritmo de 'engranajes'.
     */
    public static String descifrar(String texto_cifrado, int k) {
        if (texto_cifrado == null) return null;
        StringBuilder texto_descifrado = new StringBuilder();
        int indice_oculto = ((k % 8) + 8) % 8;

        int i = 0;
        while (i < texto_cifrado.length()) {
            char charActual = texto_cifrado.charAt(i);
            
            if (charActual == ' ') {
                texto_descifrado.append(' ');
                i++;
                continue;
            }

            if (TODOS_LOS_CARACTERES.indexOf(charActual) == -1) {
                texto_descifrado.append(charActual);
                i++;
                continue;
            }

            if (i + 8 > texto_cifrado.length()) {
                break; // Texto corrupto o bloque incompleto
            }

            String bloque = texto_cifrado.substring(i, i + 8);
            char char_cifrado = bloque.charAt(indice_oculto);

            if (ANILLO_1.indexOf(char_cifrado) != -1) {
                int c = ANILLO_1.indexOf(char_cifrado);
                int p = ((c - k) % 26 + 26) % 26;
                texto_descifrado.append(ANILLO_1.charAt(p));
            } else if (ANILLO_2.indexOf(char_cifrado) != -1) {
                int c = ANILLO_2.indexOf(char_cifrado);
                int p = ((c + k) % 26 + 26) % 26;
                texto_descifrado.append(ANILLO_2.charAt(p));
            } else if (ANILLO_3.indexOf(char_cifrado) != -1) {
                int c = ANILLO_3.indexOf(char_cifrado);
                int p = ((c - k) % 26 + 26) % 26;
                texto_descifrado.append(ANILLO_3.charAt(p));
            }

            i += 8;
        }
        return texto_descifrado.toString();
    }
}
